### PR TITLE
test: Project bootstrap tests that verifies new user experience.

### PR DESCRIPTION
**Pull request by Isakdl**
# Adds

Platform agnostic bootstrap tests that creates a new project, verifies files, and starts the server (migrations should be applied).

Related: https://github.com/serverpod/serverpod/issues/1117

## Pre-launch Checklist

- [ ] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [ ] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [ ] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [ ] I added new tests to check the change I am making.
- [ ] All existing and new tests are passing.
- [ ] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

_If you have done any breaking changes, make sure to outline them here, so that they can be included in the notes for the next release._

### CODE REVIEW

**Comment by SandPod** [comment]
This issue is also related: https://github.com/serverpod/serverpod/issues/1208

File: tests/serverpod_test_server/pubspec.yaml [hunk 1]
```diff
@@ -6,7 +6,7 @@ description: Test server for Serverpod.
 repository: https://github.com/serverpod/serverpod
 
 environment:
-  sdk: '>=2.19.0 <4.0.0'
```
Comment (SandPod) [inline tests/serverpod_test_server/pubspec.yaml:9]: Note (No action required): Should probably be a separate commit/PR.

File: tests/bootstrap_project/test/project_test.dart [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+import 'package:test/test.dart';
+import 'package:path/path.dart' as path;
+
+void main() async {
+  Directory.current = path.join(Directory.current.path, '..', '..');
+  final rootPath = Directory.current.path;
+  final cliPath = path.join(rootPath, 'tools', 'serverpod_cli');
+  final tempPath = path.join(rootPath, 'temp');
+
+  final timeout = Timeout(Duration(minutes: 5));
```
Comment (SandPod) [inline tests/bootstrap_project/test/project_test.dart:14]: NIT(Non blocking) - Use the `@Timeout` annotation to apply the same timeout to all tests in the file.

```suggestion
@Timeout(Duration(minutes: 5))

import 'dart:convert';
import 'dart:io';

import 'package:serverpod/serverpod.dart';
import 'package:test/test.dart';
import 'package:path/path.dart' as path;

void main() async {
  Directory.current = path.join(Directory.current.path, '..', '..');
  final rootPath = Directory.current.path;
  final cliPath = path.join(rootPath, 'tools', 'serverpod_cli');
  final tempPath = path.join(rootPath, 'temp');
```

File: tests/bootstrap_project/test/project_test.dart [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+
+  setUpAll(() async {
+    await Process.run(
+      'dart',
+      ['pub', 'global', 'activate', '-s', 'path', '.'],
+      workingDirectory: cliPath,
+    );
```
Comment (SandPod) [inline tests/bootstrap_project/test/project_test.dart:21]: Thought (Does not require action) - This will change the path for serverpod for the user of the host, having unexpected behavior if run locally.

I wonder if we could spawn a single shell process that we interact with to run commands. In case we could get away with only locally changing the path to the CLI.

Not sure how this would work with awaiting command executions and getting the exit code of the previous executed process and such.

File: tests/bootstrap_project/test/project_test.dart [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+        environment: {
+          'SERVERPOD_HOME': rootPath,
+        },
+      );
+
+      createProcess.stdout.transform(Utf8Decoder()).listen(print);
+      createProcess.stderr.transform(Utf8Decoder()).listen(print);
+
+      var createProjectExitCode = await createProcess.exitCode;
+      expect(createProjectExitCode, 0);
```
Comment (SandPod) [inline tests/bootstrap_project/test/project_test.dart:58]: Suggestion: add a reason so that we understand the failure in the CI messages.

```suggestion
      expect(createProjectExitCode, 0, reason: 'Should have successfully created project but exit code was != 0',);
```

File: tests/bootstrap_project/test/project_test.dart [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+
+      final docker = await Process.start(
+        'docker',
+        ['compose', 'up', '--build', '--detach'],
```
Comment (SandPod) [inline tests/bootstrap_project/test/project_test.dart:62]: Question - When we run this with `--detach`, isn't this a high risk of creating a Zoombie process? Wouldn't it be better to run this command asynchronously and close it once the test is finished?

Comment (SandPod) [reply tests/bootstrap_project/test/project_test.dart:62]: Since we retry until the database is up when we start the server in maintenance mode this should be OK as well.

File: tests/bootstrap_project/test/project_test.dart [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+      await Process.run(
+        'docker',
+        ['compose', 'down', '-v'],
+        workingDirectory: commandRootPath,
+      );
+    });
+
+    group('when creating a new project', () {
+      setUpAll(() async {
+        await Process.run(
```
Comment (SandPod) [inline tests/bootstrap_project/test/project_test.dart:105]: Suggestion - Assert the exit code is != 0 in the setup.

This will help us identify when the pre-requisite for the test is not fulfilled.

File: tests/bootstrap_project/test/project_test.dart [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+  }, timeout: timeout);
+
+  group('Given a clean state', () {
+    final (projectName, commandRoot) = createRandomProjectName(tempPath);
+    final (serverDir, _, clientDir) = createProjectFolderPaths(projectName);
+
+    late Process createProcess;
+
+    tearDown(() async {
+      print('----------------------------------------------------------');
```
Comment (SandPod) [inline tests/bootstrap_project/test/project_test.dart:325]: Question - Should this print be removed?

File: tests/bootstrap_project/test/project_test.dart [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+
+      createProcess.stdout.transform(Utf8Decoder()).listen(print);
+      createProcess.stderr.transform(Utf8Decoder()).listen(print);
+
+      var createProjectExitCode = await createProcess.exitCode;
+      expect(createProjectExitCode, 0);
+
+      // Delete generated files
+      await Process.run('rm', ['-f', '${serverDir}/lib/src/generated/*.dart']);
+      await Process.run('rm', ['-f', '${clientDir}/lib/src/protocol/*.dart']);
```
Comment (SandPod) [inline tests/bootstrap_project/test/project_test.dart:354]: Note - We should add the yaml endpoint file we accidentally removed to this test in that PR.

**Review by SandPod — COMMENTED** [review]
Nice work! Awesome to make the test platform agnostic and run through dart 🙌 

A couple of comments and suggestion.
Let me know if anything is unclear.

File: tools/serverpod_cli/lib/src/util/command_line_tools.dart [hunk 9]
```diff
@@ -209,6 +185,7 @@ Future<int> _runProcessWithDefaultLogger({
     executable,
     arguments ?? [],
     workingDirectory: workingDirectory,
+    runInShell: true,
```
[the remark this answers is not in the document]
Comment (Isakdl) [reply tools/serverpod_cli/lib/src/util/command_line_tools.dart:188]: Not anything visible as far as I have seen.

File: tests/bootstrap_project/test/project_test.dart (hunk 7 above)
Comment (Isakdl) [reply tests/bootstrap_project/test/project_test.dart:325]: yes

File: tests/bootstrap_project/test/project_test.dart (hunk 8 above)
Comment (Isakdl) [reply tests/bootstrap_project/test/project_test.dart:354]: indeed but need to have the change on main!

File: tests/bootstrap_project/test/project_test.dart (hunk 5 above)
Comment (Isakdl) [reply tests/bootstrap_project/test/project_test.dart:62]: The reason I did this was to ensure the container has started before we run our command. But I think the risk of a zombie is just as big without using --detach as we would have to kill the process manually or it will continue to run in the background. 

Now the tearDown will kill the containers.

File: tests/bootstrap_project/test/project_test.dart (hunk 3 above)
Comment (Isakdl) [reply tests/bootstrap_project/test/project_test.dart:21]: Yeah this is a drawback with this approach, I think your solution might be impossible to do inside the test, not sure. But you could make an argument that the test shouldn't manipulate the path at all. 

But then we would have to add this in the CI script and you would have to manually set this if running the test locally. However, that might be a good thing it means you can test whatever CI command you want. 

I'm not set on what is the best solution.

File: tests/bootstrap_project/test/project_test.dart (hunk 8 above)
Comment (SandPod) [reply tests/bootstrap_project/test/project_test.dart:361]: You merged it into main now right? So now you can add it to this test :)

PR that was merged: https://github.com/serverpod/serverpod/commit/885b2c2a625b223ba61f51d65434df9e917d15fd

**Review by SandPod — APPROVED** [review]
LGTM! 👍
